File: internal/service/delta.go
package service

import (
	"fmt"
	"time"
)

func annotateOverviewDelta(previous Snapshot, next Snapshot, hasPrevious bool) Snapshot {
	if !hasPrevious {
		next.Overview.Changes = OverviewDelta{}
		return next
	}
	next.Overview.Changes = buildOverviewDelta(previous, next)
	return next
}

func buildOverviewDelta(previous Snapshot, next Snapshot) OverviewDelta {
	previousSessions := map[string]SessionView{}
	for _, session := range previous.Sessions {
		previousSessions[session.Key] = session
	}
	previousTurns := map[string]struct{}{}
	for _, turn := range previous.Turns {
		previousTurns[turnKey(turn)] = struct{}{}
	}
	sessionsWithTurns := map[string]struct{}{}
	delta := OverviewDelta{}
	for _, session := range next.Sessions {
		if _, ok := previousSessions[session.Key]; !ok {
			delta.NewSessions++
		}
		if isAlertingSession(session) && !isAlertingSession(previousSessions[session.Key]) {
			delta.NewAlerts++
		}
	}
	for _, turn := range next.Turns {
		if _, ok := previousTurns[turnKey(turn)]; ok {
			continue
		}
		delta.TokensAdded += turn.TotalTokens
		if turn.SessionKey != "" {
			sessionsWithTurns[turn.SessionKey] = struct{}{}
		}
	}
	delta.SessionsWithNewTurns = len(sessionsWithTurns)

	previousAgents := map[string]struct{}{}
	for _, agent := range previous.Agents {
		previousAgents[agentKey(agent)] = struct{}{}
	}
	nextAgents := map[string]struct{}{}
	for _, agent := range next.Agents {
		key := agentKey(agent)
		nextAgents[key] = struct{}{}
		if _, ok := previousAgents[key]; !ok {
			delta.AgentsStarted++
		}
	}
	for key := range previousAgents {
		if _, ok := nextAgents[key]; !ok {
			delta.AgentsEnded++
		}
	}
	previousSourceErrors := map[string]string{}
	for _, source := range previous.Overview.Sources {
		if source.Error != "" {
			previousSourceErrors[source.Provider] = source.Error
		}
	}
	for _, source := range next.Overview.Sources {
		if source.Error != "" && previousSourceErrors[source.Provider] != source.Error {
			delta.SourceErrors++
		}
	}
	return delta
}

func isAlertingSession(session SessionView) bool {
	return session.UsageState == "warn" || session.UsageState == "stop" || session.State == "warn" || session.State == "stop"
}

func turnKey(turn TurnView) string {
	if turn.ID != "" {
		return turn.Provider + ":" + turn.SessionKey + ":" + turn.ID
	}
	if turn.RequestID != "" {
		return turn.Provider + ":" + turn.SessionKey + ":" + turn.RequestID
	}
	return fmt.Sprintf("%s:%s:%s:%s:%d:%d", turn.Provider, turn.SessionKey, turn.Model, turn.At.UTC().Format(time.RFC3339Nano), turn.TotalTokens, turn.CumulativeTokens)
}

func agentKey(agent AgentView) string {
	started := ""
	if agent.ProcessStartedAt != nil {
		started = agent.ProcessStartedAt.UTC().Format(time.RFC3339Nano)
	}
	return fmt.Sprintf("%s:%d:%s", agent.ID, agent.PID, started)
}

File: curb-core/src/service/delta.rs
use std::collections::{BTreeSet, HashMap};

use super::{AgentView, OverviewDelta, SessionView, Snapshot, TurnView};

pub fn annotate_overview_delta(previous: Option<&Snapshot>, mut next: Snapshot) -> Snapshot {
    next.overview.changes = previous
        .map(|previous| build_overview_delta(previous, &next))
        .unwrap_or_default();
    next
}

fn build_overview_delta(previous: &Snapshot, next: &Snapshot) -> OverviewDelta {
    let previous_sessions = previous
        .sessions
        .iter()
        .map(|session| (session.key.as_str(), session))
        .collect::<HashMap<_, _>>();
    let previous_turns = previous.turns.iter().map(turn_key).collect::<BTreeSet<_>>();
    let mut sessions_with_turns = BTreeSet::new();
    let mut delta = OverviewDelta::default();
    for session in &next.sessions {
        let previous = previous_sessions.get(session.key.as_str()).copied();
        if previous.is_none() {
            delta.new_sessions += 1;
        }
        if is_alerting_session(session) && !previous.is_some_and(is_alerting_session) {
            delta.new_alerts += 1;
        }
    }
    for turn in &next.turns {
        if previous_turns.contains(&turn_key(turn)) {
            continue;
        }
        delta.tokens_added += turn.spent_tokens;
        if !turn.session_key.is_empty() {
            sessions_with_turns.insert(turn.session_key.clone());
        }
    }
    delta.sessions_with_new_turns = sessions_with_turns.len();

    let previous_agents = previous
        .agents
        .iter()
        .map(agent_key)
        .collect::<BTreeSet<_>>();
    let next_agents = next.agents.iter().map(agent_key).collect::<BTreeSet<_>>();
    delta.agents_started = next_agents.difference(&previous_agents).count();
    delta.agents_ended = previous_agents.difference(&next_agents).count();

    let previous_source_errors = previous
        .overview
        .sources
        .iter()
        .filter_map(|source| {
            source
                .error
                .as_ref()
                .map(|error| (source.provider.as_str(), error.as_str()))
        })
        .collect::<HashMap<_, _>>();
    delta.source_errors = next
        .overview
        .sources
        .iter()
        .filter(|source| {
            source.error.as_ref().is_some_and(|error| {
                previous_source_errors
                    .get(source.provider.as_str())
                    .is_none_or(|previous| previous != error)
            })
        })
        .count();
    delta
}

fn is_alerting_session(session: &SessionView) -> bool {
    session.alert != "ok"
}

fn turn_key(turn: &TurnView) -> String {
    if let Some(id) = &turn.id
        && !id.is_empty()
    {
        return format!("{}:{}:{}", turn.provider, turn.session_key, id);
    }
    if let Some(request_id) = &turn.request_id
        && !request_id.is_empty()
    {
        return format!("{}:{}:{}", turn.provider, turn.session_key, request_id);
    }
    format!(
        "{}:{}:{}:{}:{}:{}",
        turn.provider,
        turn.session_key,
        turn.model.as_deref().unwrap_or_default(),
        turn.at.map(|at| at.to_rfc3339()).unwrap_or_default(),
        turn.total_tokens,
        turn.cumulative_tokens
    )
}

fn agent_key(agent: &AgentView) -> String {
    format!(
        "{}:{}:{}",
        agent.id,
        agent.pid,
        agent
            .process_started_at
            .map(|started_at| started_at.to_rfc3339())
            .unwrap_or_default()
    )
}
